tests/compiletest.rs

```diff
@@ -1,71 +1,432 @@
+
+    let failures = failures.into_inner().unwrap();
+    let total = total.load(Ordering::Relaxed);
+    let skipped = skipped.load(Ordering::Relaxed);
+    if !failures.is_empty() {
+        for (path, miri, revision, errors) in &failures {
+            eprintln!();
+            eprint!("{} {}", path.display().to_string().underline(), "FAILED".red());
+            if !revision.is_empty() {
+                eprint!(" (revision `{}`) ", revision);
+            }
+            eprintln!();
+            eprintln!("command: {:?}", miri);
+            for error in errors {
+                match error {
+                    Error::ExitStatus(mode, exit_status) => eprintln!("{mode:?} got {exit_status}"),
+                    Error::PatternNotFound(pat) =>
+                        eprintln!("error-pattern `{pat}` not found in stderr output"),
+                    Error::InlinePatternNotFound(pat) =>
+                        eprintln!("`{pat}` not found in stderr output"),
+                    Error::NoPatternsFound => eprintln!("no error patterns found in failure test"),
+                    Error::OutputDiffers { path, actual, expected } =>
+                        compare_output(path, actual, expected),
+                }
+            }
+        }
+        eprintln!();
+        eprintln!(
+            "{} tests failed, {} tests passed, {} skipped",
+            failures.len().to_string().red().bold(),
+            (total - failures.len() - skipped).to_string().green(),
+            skipped.to_string().yellow()
+        );
+        std::process::exit(1);
+    }
+    eprintln!();
     eprintln!(
-        "{}",
-        format!("## Running compile-fail tests in {} against miri for target {}", path, target)
-            .green()
-            .bold()
+        "{} tests passed, {} skipped",
+        (total - skipped).to_string().green(),
+        skipped.to_string().yellow()
     );
+}
+
+#[derive(Debug)]
+enum Error {
+    /// Got an invalid exit status for the given mode.
+    ExitStatus(Mode, ExitStatus),
+    PatternNotFound(String),
+    InlinePatternNotFound(String),
+    NoPatternsFound,
+    OutputDiffers {
+        path: PathBuf,
+        actual: String,
+        expected: String,
+    },
+}
+
+type Errors = Vec<Error>;
+
+fn revisions(path: &Path) -> Vec<String> {
+    let content = std::fs::read_to_string(path).unwrap();
+    for line in content.lines() {
+        if let Some(revisions) = line.strip_prefix("// revisions:") {
+            return revisions.trim().split_whitespace().map(|s| s.to_string()).collect();
+        }
+    }
+    vec![String::new()]
+}
+
+fn run_test(
+    path: &Path,
+    target: &str,
+    flags: &[String],
+    mode: Mode,
+    revision: &str,
+) -> (Command, Errors) {
+    // Run miri
+    let mut miri = Command::new(miri_path());
+    miri.args(flags.iter());
+    miri.arg(path);
+    if !revision.is_empty() {
+        miri.arg(format!("--cfg={revision}"));
+    }
+    miri.env("RUSTC_BACKTRACE", "0");
```

Comment:
Why this?

Reply:
This is what happened before. Otherwise locally setting the env var influences ui test output

Reply:
Happened before where? In compiletest?
```suggestion
    // Tests should work even when people have that env var set.
    miri.env("RUSTC_BACKTRACE", "0");
```